File: src/Find_PlantStableRegion_MonitoringOneHDV.m
%% Description
% Find the plant stable region when monitoring one HDV

clc;
clear;
close all;

%% Parameters


% number of the following HDVs
n = 2;
% number of the preceding HDVs
m = 2;

DriverDynamics = 1;






if m ~= 0 && n ~= 0
    ID = zeros(1,m+n);
    ID(1:m) = -m:-1;
    ID(m+1:m+n) = 1:n;
elseif m == 0
    ID = 1:n;
elseif n == 0
    ID = -m:-1;
end

% id of the HDV that has feedback
for id = ID
    
    K = -10:0.1:10;
    Mu = -10:0.1:10;
    
    s_star = 20;
    
    
    %%
    
    %%%%% Please Change %%%%%
    switch DriverDynamics
        case 1
            alpha = 0.6;
            beta = 0.9;
        case 2
            alpha = 0.4;
            beta = 0.6;
    end
    
    %%%%% No change %%%%%
    v_max  = 30;
    s_st   = 5;
    s_go   = 35;
    alpha1 = alpha*v_max/2*pi/(s_go-s_st)*sin(pi*(s_star-s_st)/(s_go-s_st));
    alpha2 = alpha+beta;
    alpha3 = beta;
    
    [A,B] = SystemModel_GeneralLCC(n,m,alpha1,alpha2,alpha3);
    
    PS_bool = zeros(length(K),length(Mu));
    
    %%
    h=waitbar(0,'please wait');
    iTest = 1;
    TestNumber = length(K)*length(Mu);
    
    for ik0 = 1:length(K)
        for ik1 = 1:length(Mu)
            k = K(ik0);
            mu = Mu(ik1);
            
            Feedback = zeros(1,2*(m+n+1));
            
            Feedback(2*m+1) = alpha1;
            Feedback(2*m+2) = -alpha2;
            Feedback(2*m) = alpha3;
            
            Feedback(2*(m+id)+1) = Feedback(2*(m+id)+1)+k;
            Feedback(2*(m+id)+2) = Feedback(2*(m+id)+2)+mu;
            
            A_closed = A+B*Feedback;
            
            
            
            
            if isempty(find(real(eig(A_closed))>0,1))
                
                PS_bool(ik0,ik1) = 1;
                
            end
            str=['DriverDynamics=',num2str(DriverDynamics),'; ID=',num2str(id),...
                '; Processing...',num2str(iTest/TestNumber*100),'%'];
            waitbar(iTest/TestNumber,h,str);
            iTest = iTest+1;
        end
    end
    
    close(h);
    
%     save(['..\_data\',date,'_PSRegion_n_',num2str(n),'_m_',num2str(m),...
%         '_DriverDynamics_',num2str(DriverDynamics),'_FeedbackID_',num2str(id),'.mat']);
    
end

File: Python_Implementation/src/Find_PlantStableRegion_MonitoringOneHDV.py
import numpy as np
from scipy.linalg import eig
import os
from tqdm import tqdm 
from datetime import date
import datetime
import scipy.io as sio
from SystemModel_GeneralLCC import SystemModel_GeneralLCC

## Description
# Find the plant stable region when monitoring one HDV

## Parameters
# number of the following HDVs
n = 2  
# number of the preceding HDVs
m = 2  

DriverDynamics = 1

if m != 0 and n != 0:
    ID = np.zeros(m + n, dtype=int)
    ID[:m] = np.arange(-m, 0, 1)
    ID[m:] = np.arange(1, n + 1, 1)
elif m == 0:
    ID = np.arange(1, n + 1, 1)
elif n == 0:
    ID = np.arange(-m, 0, 1)

# id of the HDV that has feedback
for id in ID:
    K = np.arange(-10, 10.1, 0.1)
    Mu = np.arange(-10, 10.1, 0.1)
    s_star = 20

    ##### Please Change #####
    if DriverDynamics == 1:
        alpha, beta = 0.6, 0.9
    elif DriverDynamics == 2:
        alpha, beta = 0.4, 0.6

    ##### No change #####
    v_max = 30
    s_st = 5
    s_go = 35
    alpha1 = alpha * v_max / 2 * np.pi / (s_go - s_st) * np.sin(np.pi * (s_star - s_st) / (s_go - s_st))
    alpha2 = alpha + beta
    alpha3 = beta

    A, B = SystemModel_GeneralLCC(n, m, alpha1, alpha2, alpha3)
    PS_bool = np.zeros((len(K), len(Mu)))

    total_tests = len(K) * len(Mu)
    pbar = tqdm(total=total_tests, desc=f'DriverDynamics={DriverDynamics}, ID={id}')

    for ik0 in range(1, len(K)+1):
        for ik1 in range(1, len(Mu)+1):
            k = K[ik0 - 1]
            mu = Mu[ik1 - 1]
            Feedback = np.zeros(2 * (m + n + 1))
            Feedback[2 * m ] = alpha1
            Feedback[2 * m + 1] = -alpha2
            Feedback[2 * m - 1] = alpha3

            Feedback[2 * (m + id)] += k
            Feedback[2 * (m + id) + 1] += mu

            Feedback = Feedback.reshape(1, 10)
            A_closed = A + np.dot(B, Feedback)

            if not np.any(np.real(np.linalg.eigvals(A_closed)) > 0):
                PS_bool[ik0 - 1, ik1 - 1] = 1
            
            pbar.update(1)
    pbar.close()

    home_dir = os.path.expanduser("~")
    data_dir = os.path.join(home_dir, "LCC", "Python_Implementation", "_data")

    #np.save(os.path.join(data_dir, f"{date.today()}_PSRegion_n_{n}_m_{m}_DriverDynamics_{DriverDynamics}_FeedbackID_{id}.npy"), PS_bool)
    filename =  datetime.datetime.now().strftime('%Y%m%d') 
    filename += "_PSRegion_n_{}".format(n)
    filename += "_m_{}".format(m)
    filename += "_DriverDynamics_{}".format(DriverDynamics)
    filename += "_FeedbackID_{}.mat".format(id)
    data_to_save = {
        "id": id,
        "DriverDynamics": DriverDynamics,
        "n": n,
        "m": m,
        "K": K,
        "Mu": Mu,
        "s_star": s_star,
        "alpha": alpha,
        "beta": beta,
        "v_max": v_max,
        "s_st": s_st,
        "s_go": s_go,
        "alpha1": alpha1,
        "alpha2": alpha2,
        "alpha3": alpha3,
        "PS_bool": PS_bool
    }
    full_path = os.path.join(data_dir, filename)
    sio.savemat(full_path, data_to_save)
